### PR TITLE
Add gpt-neo-x support

### LINKED ISSUE #555: Community contribution -  `optimum.exporters.onnx` support for new models!
Following what was done by @ChainYo in Transformers, in the [ONNXConfig: Add a configuration for all available models](https://github.com/huggingface/transformers/issues/16308) issue, the idea is to add support for exporting new models in `optimum.exporters.onnx`.

This issue is about the working group specially created for this task. If you are interested in helping out, reply here, take a look at this [organization](https://huggingface.co/OWG), or add  `ChainYo#3610` on discord.

We want to contribute to Hugging Face's ONNX export implementation for all available models on Hugging Face Hub. There are already a lot of architectures implemented for converting PyTorch models to ONNX, but we need more! We need them all!

Feel free to join us in this adventure! Join the org by clicking [here](https://huggingface.co/organizations/OWG/share/TskjfGaGjGnMXXssbPPXrQWEIbosGqZshZ)

Here is a non-exhaustive list of models that all models available:

- [x] Albert
- [x] BART
- [x] BeiT
- [x] BERT
- [ ] BigBird (Critical issue: https://github.com/huggingface/optimum/issues/754#issuecomment-1429830467)
- [ ] BigBirdPegasus (Critical issue: https://github.com/huggingface/optimum/issues/754#issuecomment-1429830467)
- [x] Blenderbot
- [x] BlenderbotSmall
- [ ] BLIP-2
- [x] BLOOM
- [x] CamemBERT
- [ ] CANINE
- [x] CLIP
- [x] CodeGen
- [x] ConvNext
- [x] ConvBert
- [ ] CTRL
- [x] CvT
- [x] Data2VecText
- [x] Data2VecVision
- [x] Deberta
- [x] DebertaV2
- [x] DeiT
- [ ] DecisionTransformer
- [x] DETR
- [x] Distilbert
- [ ] DPR
- [x] DPT
- [x] ELECTRA
- [ ] FNet
- [ ] FSMT
- [x] Flaubert
- [ ] FLAVA
- [ ] Funnel Transformer
- [x] GLPN
- [x] GPT2
- [x] GPTJ
- [x] GPT-Neo
- [x] GPT-NeoX
- [x] Hubert
- [x] I-Bert
- [x] ImageGPT 🛠️ @adit299 
- [ ] LED
- [x] LayoutLM
- [ ] LayoutLMv2 (but 🛠️ in Transformers)
- [x] LayoutLMv3
- [ ] LayoutXLM
- [ ] LED
- [x] LeViT
- [ ] 🛠️ Longformer (Critical issue: https://github.com/huggingface/optimum/issues/776#issuecomment-1429680121)
- [x] LongT5
- [ ] Luke (but 🛠️ in Transformers)
- [ ] Lxmert
- [x] M2M100
- [ ] MaskFormer
- [x] mBart
- [ ] MCTCT
- [x] MPNet
- [x] MT5
- [x] MarianMT
- [ ] MegatronBert
- [x] MobileBert
- [x] MobileViT
- [ ] Nyströmformer
- [x] OpenAIGPT-2
- [x] OPT ((but 🛠️ in Transformers)
- [x] OWLViT
- [x] Pix2Struct
- [x] PLBart
- [x] Pegasus
- [x] Perceiver
- [x] PoolFormer
- [ ] ProphetNet
- [ ] QDQBERT
- [ ] RAG
- [ ] REALM
- [ ] Reformer (but 🛠️ in Transformers)
- [x] RemBert
- [x] ResNet
- [x] RegNet 🛠️ @asrimanth 
- [ ] RetriBert
- [x] RoFormer
- [x] RoBERTa
- [x] SEW
- [x] SEW-D
- [x] SegFormer
- [x] Speech2Text
- [ ] Speech2Text2
- [x] Splinter
- [x] SqueezeBERT
- [x] Swin Transformer
- [x] T5
- [ ] TAPAS 🛠️ @someshfengde 
- [ ] TAPEX
- [ ] Transformer XL
- [x] TrOCR
- [ ] UniSpeech
- [ ] UniSpeech-SAT
- [ ] VAN
- [x] ViT
- [ ] Vilt
- [ ] VisualBERT
- [x] Wav2Vec2
- [x] WavLM
- [x] Whisper
- [ ] XGLM
- [x] XLM
- [ ] XLMProphetNet
- [x] XLM-RoBERTa
- [x] XLM-RoBERTa-XL
- [ ]  XLNet (but 🛠️ in Transformers)
- [x] YOLOS
- [ ] Yoso

🛠️ next to a model suggests that the PR is in progress. If there is nothing next to a model, it means that ONNX does not yet support the model, and thus we need to add support for it.

If you need help implementing an unsupported model, here is a [guide](https://huggingface.co/docs/optimum/main/en/exporters/onnx/usage_guides/contribute) from HuggingFace Optimum documentation.

### PR DESCRIPTION
**Pull request by sidthekidder**
# What does this PR do?
Support GPT-Neo-X model config. 

Fixes #555 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sidthekidder** [comment]
Hi @fxmarty, I'm trying to add support for GPT-Neo-X model (similar to the existing GPT-Neo model). Though I'm not sure what's the best way to test this locally?

File: optimum/exporters/onnx/model_configs.py [hunk 1]
```diff
@@ -192,6 +192,11 @@ class GPTNeoOnnxConfig(TextDecoderOnnxConfig):
     NORMALIZED_CONFIG_CLASS = NormalizedTextConfig.with_args(num_attention_heads="num_heads")
 
 
+class GPTNeoXOnnxConfig(TextDecoderOnnxConfig):
+    DEFAULT_ONNX_OPSET = 13
+    NORMALIZED_CONFIG_CLASS = NormalizedTextConfig.with_args(num_attention_heads="num_attention_heads")
```
Comment (fxmarty) [inline optimum/exporters/onnx/model_configs.py:197]:
```suggestion
    NORMALIZED_CONFIG_CLASS = NormalizedTextConfig
```

It is enough as `num_attention_heads` is the default: https://github.com/huggingface/optimum/blob/40a01b3c883ca3c092a4493d3f5ca524ed3109ab/optimum/utils/normalized_config.py#L72

File: optimum/exporters/tasks.py [hunk 2]
```diff
@@ -382,6 +382,14 @@ class TasksManager:
             "sequence-classification",
             onnx="GPTNeoOnnxConfig",
         ),
+        "gpt-neo-x": supported_tasks_mapping(
+            "default",
+            "default-with-past",
+            "causal-lm",
+            "causal-lm-with-past",
+            "sequence-classification",
```
Comment (fxmarty) [inline optimum/exporters/tasks.py:390]: Does it really support it?

File: optimum/exporters/tasks.py [hunk 3]
```diff
@@ -382,6 +382,14 @@ class TasksManager:
             "sequence-classification",
             onnx="GPTNeoOnnxConfig",
         ),
+        "gpt-neo-x": supported_tasks_mapping(
+            "default",
+            "default-with-past",
```
Comment (fxmarty) [inline optimum/exporters/tasks.py:387]:
```suggestion
```

File: optimum/utils/normalized_config.py [hunk 4]
```diff
@@ -187,6 +187,7 @@ class NormalizedConfigManager:
         "electra": NormalizedTextConfig,
         "gpt2": GPT2LikeNormalizedTextConfig,
         "gpt_neo": NormalizedTextConfig.with_args(num_attention_heads="num_heads"),
+        "gpt_neo_x": NormalizedTextConfig.with_args(num_attention_heads="num_attention_heads"),
```
Comment (fxmarty) [inline optimum/utils/normalized_config.py:190]:
```suggestion
        "gpt_neo_x": NormalizedTextConfig,
```

File: tests/onnxruntime/test_modeling.py [hunk 5]
```diff
@@ -108,6 +108,7 @@
     "flaubert": "hf-internal-testing/tiny-random-flaubert",
     "gpt2": "hf-internal-testing/tiny-random-gpt2",
     "gpt_neo": "hf-internal-testing/tiny-random-GPTNeoModel",
+    "gpt_neo_x": "hf-internal-testing/tiny-random-GPTNeoXForCausalLM",
```
Comment (fxmarty) [inline tests/onnxruntime/test_modeling.py:111]: Could you add `gpt_neo_x` in the `ORTModelForCausalLMIntegrationTest`'s `SUPPORTED_ARCHITECTURES`? This allows to test it using the ONNX Runtime integration.

**Review by fxmarty — COMMENTED** [review]
Looks great, thank you for your contribution! Left a few comments.

File: optimum/exporters/tasks.py (hunk 3 above)
Comment (fxmarty) [reply optimum/exporters/tasks.py:387]: Apologies, `default-with-past` is actually a thing.

File: optimum/exporters/onnx/model_configs.py (hunk 1 above)
Comment (sidthekidder) [reply optimum/exporters/onnx/model_configs.py]: Makes sense, updated!

File: tests/onnxruntime/test_modeling.py (hunk 5 above)
Comment (sidthekidder) [reply tests/onnxruntime/test_modeling.py:111]: Added!

**Review by fxmarty — APPROVED** [review]
It is all good, thank you for the addition!

**Pull request merged**